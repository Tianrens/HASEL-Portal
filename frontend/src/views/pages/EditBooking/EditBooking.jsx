import { React, useEffect, useState } from 'react';
import { useParams, useHistory } from 'react-router-dom';
import { useSnackbar } from 'notistack';
import TopBarPageTemplate from '../../components/templates/TopBarPageTemplate/TopBarPageTemplate';
import { useCrud } from '../../../hooks/useCrud';
import { header } from './EditBooking.module.scss';
import BottomButtons from '../../components/buttons/BottomButtons';
import { authRequest } from '../../../hooks/util/authRequest';
import BookingForm from '../../components/forms/BookingForm';

const EditBooking = () => {
    const [bookingState, setBookingState] = useState({});
    const [error, setError] = useState(true);
    const [userWorkstation, setUserWorkstation] = useState(null);

    const userWorkstationName = userWorkstation?.name;
    const numGPUs = userWorkstation?.numGPUs;

    const { enqueueSnackbar, closeSnackbar } = useSnackbar();
    const history = useHistory();
    const successCallback = (message) => {
        enqueueSnackbar(message, {
            variant: 'success',
            autoHideDuration: 3000,
            onClose: closeSnackbar,
        });
        history.goBack();
    };

    const errorCallback = (message) => {
        enqueueSnackbar(message, {
            variant: 'error',
            autoHideDuration: 3000,
            onClose: closeSnackbar,
        });
    };

    const { bookingId } = useParams();
    const booking = useCrud(`/api/booking/${bookingId}`).data;

    const onDelete = async () => {
        try {
            await authRequest(`/api/booking/${bookingId}`, 'DELETE');
            successCallback('Booking deleted successfully');
        } catch (err) {
            errorCallback(err.message);
        }
    };

    const onCancel = () => {
        history.goBack();
    };

    const onAcceptChanges = async () => {
        if (error) {
            return;
        }
        try {
            await authRequest(`/api/booking/${bookingId}`, 'PUT', {
                workstationId: userWorkstation?._id,
                ...bookingState,
            });
            successCallback('Booking updated successfully');
        } catch (err) {
            errorCallback(err.message);
        }
    };

    useEffect(() => {
        const getAndSetValues = async () => {
            const workstationResponse = await authRequest(
                `/api/workstation/${booking ? booking?.workstationId : ''}`,
            );
            setUserWorkstation(workstationResponse.data);
        };
        getAndSetValues();
    }, [booking]);

    return (
        <TopBarPageTemplate>
            <h2 className={header}>Edit Booking - {userWorkstationName}</h2>
            {userWorkstation && (
                <BookingForm
                    updateBookingState={(newBookingState, isError) => {
                        setError(isError);
                        setBookingState(newBookingState);
                    }}
                    numGPUs={numGPUs}
                    data={booking}
                    workstationId={userWorkstation._id}
                />
            )}
            <BottomButtons
                onDelete={onDelete}
                onAccept={onAcceptChanges}
                onDeny={onCancel}
                acceptText='Confirm Changes'
                denyText='Cancel'
                acceptDisabled={error}
            />
        </TopBarPageTemplate>
    );
};

export default EditBooking;
