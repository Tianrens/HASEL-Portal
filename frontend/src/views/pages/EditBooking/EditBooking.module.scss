@import '../../../assets/responsiveDesignHelpers';
@import '../../../assets/colours';

.header {
    align-self: flex-start;
    margin-bottom: 4rem;
}
