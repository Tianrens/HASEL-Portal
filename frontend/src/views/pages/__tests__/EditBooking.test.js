import React from 'react';
import renderer from 'react-test-renderer';
import { StylesProvider } from '@material-ui/core';
import { MemoryRouter } from 'react-router-dom';
import { SnackbarProvider } from 'notistack';
import EditBooking from '../EditBooking/EditBooking';

test('Edit Booking page renders properly', () => {
    const snapshotComponent = renderer.create(
        <StylesProvider injectFirst>
            <SnackbarProvider>
                <MemoryRouter>
                    <EditBooking />
                </MemoryRouter>
            </SnackbarProvider>
        </StylesProvider>,
    );
    const tree = snapshotComponent.toJSON();
    expect(tree).toMatchSnapshot();
});
