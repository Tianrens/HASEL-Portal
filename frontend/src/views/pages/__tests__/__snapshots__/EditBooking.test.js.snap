// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Booking page renders properly 1`] = `
Array [
  <div
    className="fillTopbarSpace"
  >
    <div
      className="navbar"
    >
      <div
        className="patternBgLayer"
      >
        <div
          className="bgLayer"
        >
          <a
            href="/"
            onClick={[Function]}
          >
            <img
              alt="University of Auckland"
              className="uoaLogo"
              src="uoaLogo.svg"
            />
            <div
              className="spacer"
            />
            <div
              className="title"
            >
              HASEL Lab
            </div>
          </a>
          <div
            className="pushSpacer"
          />
          <div
            className="horizontalNavLinks"
          >
            <div
              className="linksWrapper"
            >
              <button
                className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root wrapper css-1e6y48t-MuiButtonBase-root-MuiButton-root"
                disabled={false}
                onBlur={[Function]}
                onContextMenu={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                <a
                  href="/user"
                  onClick={[Function]}
                >
                  <img
                    alt="My Account"
                    className="navIcon"
                    src="id.svg"
                  />
                  <div
                    className="spacer"
                  />
                  <div
                    className="title"
                  >
                    My Account
                  </div>
                </a>
              </button>
            </div>
          </div>
          <div
            className="drawerWrapper"
            onClick={[Function]}
            onTouchEnd={[Function]}
          >
            <button
              className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge drawerButton css-mf1cb5-MuiButtonBase-root-MuiIconButton-root"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="DehazeIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M2 15.5v2h20v-2H2zm0-5v2h20v-2H2zm0-5v2h20v-2H2z"
                />
              </svg>
            </button>
            <div
              className="MuiDrawer-root MuiDrawer-docked drawer css-7ik032-MuiDrawer-docked"
            >
              <div
                className="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiDrawer-paper MuiDrawer-paperAnchorRight MuiDrawer-paperAnchorDockedRight css-18sg6k4-MuiPaper-root-MuiDrawer-paper"
                style={
                  Object {
                    "visibility": "hidden",
                  }
                }
              >
                <div
                  className="drawerPatternBg"
                >
                  <div
                    className="drawer"
                  >
                    <div>
                      <button
                        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge drawerButton css-mf1cb5-MuiButtonBase-root-MuiIconButton-root"
                        disabled={false}
                        onBlur={[Function]}
                        onClick={[Function]}
                        onContextMenu={[Function]}
                        onDragLeave={[Function]}
                        onFocus={[Function]}
                        onKeyDown={[Function]}
                        onKeyUp={[Function]}
                        onMouseDown={[Function]}
                        onMouseLeave={[Function]}
                        onMouseUp={[Function]}
                        onTouchEnd={[Function]}
                        onTouchMove={[Function]}
                        onTouchStart={[Function]}
                        tabIndex={0}
                        type="button"
                      >
                        <svg
                          aria-hidden={true}
                          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                          data-testid="CloseIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                          />
                        </svg>
                      </button>
                    </div>
                    <div
                      className="linksWrapper"
                    >
                      <button
                        className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root wrapper css-1e6y48t-MuiButtonBase-root-MuiButton-root"
                        disabled={false}
                        onBlur={[Function]}
                        onContextMenu={[Function]}
                        onDragLeave={[Function]}
                        onFocus={[Function]}
                        onKeyDown={[Function]}
                        onKeyUp={[Function]}
                        onMouseDown={[Function]}
                        onMouseLeave={[Function]}
                        onMouseUp={[Function]}
                        onTouchEnd={[Function]}
                        onTouchMove={[Function]}
                        onTouchStart={[Function]}
                        tabIndex={0}
                        type="button"
                      >
                        <a
                          href="/user"
                          onClick={[Function]}
                        >
                          <img
                            alt="My Account"
                            className="navIcon"
                            src="id.svg"
                          />
                          <div
                            className="spacer"
                          />
                          <div
                            className="title"
                          >
                            My Account
                          </div>
                        </a>
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="content"
  >
    <h2
      className="header"
    >
      Edit Booking - 
    </h2>
    <div
      className="buttonContainer"
    >
      <div>
        <button
          className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root button yellow   css-sghohy-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="submit"
        >
          <span
            className="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
          >
            <span
              aria-hidden={true}
              className="material-icons notranslate MuiIcon-root MuiIcon-fontSizeMedium css-kp9ftd-MuiIcon-root"
            >
              delete
            </span>
          </span>
          Delete
        </button>
      </div>
      <div>
        <button
          className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root button red   css-sghohy-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="submit"
        >
          <span
            className="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
          >
            <span
              aria-hidden={true}
              className="material-icons notranslate MuiIcon-root MuiIcon-fontSizeMedium css-kp9ftd-MuiIcon-root"
            >
              close
            </span>
          </span>
          Cancel
        </button>
        <button
          className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root Mui-disabled button green   css-sghohy-MuiButtonBase-root-MuiButton-root"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          type="submit"
        >
          <span
            className="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
          >
            <span
              aria-hidden={true}
              className="material-icons notranslate MuiIcon-root MuiIcon-fontSizeMedium css-kp9ftd-MuiIcon-root"
            >
              done
            </span>
          </span>
          Confirm Changes
        </button>
      </div>
    </div>
  </div>,
]
`;
